runtime/lua/vim/_memoize.lua

```diff
@@ -0,0 +1,51 @@
+--- Module for private utility functions
+
+local function concat_hash(...)
+  return table.concat({ ... })
+end
+
+local function idx_hash(idx)
+  return function(...)
+    return select(idx, ...)
+  end
+end
+
+--- @param hash_fn any
+--- @return fun(...): any
+local function resolve_hash(hash_fn)
+  if type(hash_fn) == 'number' then
+    hash_fn = idx_hash(hash_fn)
+  elseif type(hash_fn) == 'string' then
+    if hash_fn == 'concat' then
+      hash_fn = concat_hash
+    else
+      error('invalid value for hash_fn: ' .. hash_fn)
+    end
+  end
+  return hash_fn
+end
+
+--- @generic F: function
+--- @param hash? integer|'concat'|fun(...): any
```

Comment:
Not optional anymore:
```suggestion
--- @param hash integer|'concat'|fun(...): any
```